crates/transaction-emitter-lib/src/emitter/account_minter.rs

```diff
@@ -353,15 +361,73 @@ pub async fn execute_and_wait_transactions(
         account.address()
     );
 
-    let pending_txns: Vec<Response<PendingTransaction>> = try_join_all(
-        txns.iter()
-            .map(|t| RETRY_POLICY.retry(move || client.submit(t))),
-    )
-    .await?;
+    async fn submit_batch(
+        client: &RestClient,
+        state_mutex: &Mutex<SubmittingTxnState>,
+    ) -> Result<()> {
+        let (indices, txns) = {
+            let state = state_mutex.lock();
+
+            let mut indices = Vec::new();
+            let mut txns = Vec::new();
+            for (i, txn) in state.txns.iter().enumerate() {
+                if state.results.get(i).map(|r| r.is_failure()).unwrap_or(true) {
+                    indices.push(i);
+                    txns.push(txn.clone());
+                }
+            }
+            (indices, txns)
+        };
+
+        let results = client.submit_batch_bcs(&txns).await.unwrap().into_inner();
+
+        let mut state = state_mutex.lock();
+        for (idx, result) in results.transaction_results.into_iter().enumerate() {
+            let input_idx = *indices.get(idx).unwrap();
+            if state.results.len() == input_idx {
+                // We need to fill it up on the first call:
+                state.results.push(result);
+            } else {
+                state.results[input_idx] = result;
+            }
+        }
+
+        if state.results.iter().any(|r| r.is_failure()) {
+            Err(format_err!(""))
+        } else {
+            Ok(())
+        }
+    }
+
+    let state_mutex = Mutex::new(SubmittingTxnState {
```

Comment:
Can you explain why a Mutex is needed here? This should all be done in one thread right?